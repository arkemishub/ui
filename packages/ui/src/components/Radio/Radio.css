/**
 * Copyright 2023 Arkemis S.r.l.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

.radio {
  @apply border rounded-full h-4 w-4 relative flex justify-center items-center;
}

.radio--primary {
  @apply border-primary;
}

.radio--secondary {
  @apply border-secondary;
}

.radio--primary--checked {
  @apply bg-primary;
}

.radio--secondary--checked {
  @apply bg-secondary;
}
