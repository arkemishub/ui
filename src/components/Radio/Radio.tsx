/**
 * Copyright 2023 Arkemis S.r.l.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from "react";
import { IRadioProps } from "./Radio.types";
import { twMerge } from "tailwind-merge";

function Radio({
  label,
  name,
  labelPlacement = "right",
  value,
  checked = false,
  onChange,
  color = "secondary",
  disabled,
}: IRadioProps) {
  return (
    <div
      className={twMerge(
        "radio__container",
        "w-fit",
        labelPlacement === "top" && "flex flex-col gap-1",
        labelPlacement === "bottom" && "flex flex-col-reverse gap-1",
        labelPlacement === "right" &&
          "flex flex-row-reverse items-center gap-2",
        labelPlacement === "left" && "flex flex-row items-center gap-2"
      )}
    >
      {label && <label htmlFor={name}>{label}</label>}
      <div
        data-testid="arke-radio"
        className={twMerge(
          "radio",
          color === "primary" && "radio--primary",
          color === "secondary" && "radio--secondary"
        )}
      >
        <input
          className="opacity-0"
          type="radio"
          name={name}
          value={value}
          checked={checked}
          onChange={onChange}
          disabled={disabled}
        />
        {checked && !disabled && (
          <span
            data-testid="arke-radio-tick"
            className={twMerge(
              "h-2.5 w-2.5 absolute rounded-full",
              color === "primary" && "radio--primary--checked",
              color === "secondary" && "radio--secondary--checked"
            )}
          />
        )}
      </div>
    </div>
  );
}

export default Radio;
