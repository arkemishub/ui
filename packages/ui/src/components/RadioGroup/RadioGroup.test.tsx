/**
 * Copyright 2023 Arkemis S.r.l.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { RadioGroup } from "./index";
import { act, render } from "@testing-library/react";
import { Radio } from "../Radio";

const mockOptions = [
  { label: "Option 1", value: "option1" },
  { label: "Option 2", value: "option2" },
  { label: "Option 3", value: "option3" },
];

describe("RadioGroup", () => {
  test("should match snapshot", () => {
    const { asFragment } = render(
      <RadioGroup name="test" onChange={() => null}>
        {mockOptions.map((option, index) => (
          <Radio key={index} value={option.value} />
        ))}
      </RadioGroup>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("should render radio group", () => {
    const { getByTestId } = render(
      <RadioGroup name="test" onChange={() => null}>
        {mockOptions.map((option, index) => (
          <Radio key={index} value={option.value} />
        ))}
      </RadioGroup>
    );
    expect(getByTestId("arke-radio-group")).toBeInTheDocument();
  });

  test("should render all radio buttons", () => {
    const { getAllByTestId } = render(
      <RadioGroup name="test" onChange={() => null}>
        {mockOptions.map((option, index) => (
          <Radio key={index} value={option.value} />
        ))}
      </RadioGroup>
    );
    expect(getAllByTestId("arke-radio")).toHaveLength(3);
  });

  test("should call onChange when radio is clicked", () => {
    const onChange = jest.fn();
    const { getAllByRole } = render(
      <RadioGroup name="test" onChange={onChange}>
        {mockOptions.map((option, index) => (
          <Radio key={index} value={option.value} />
        ))}
      </RadioGroup>
    );

    const radio = getAllByRole("radio")[0];
    act(() => {
      radio.click();
    });

    expect(onChange).toHaveBeenCalled();
  });
});
