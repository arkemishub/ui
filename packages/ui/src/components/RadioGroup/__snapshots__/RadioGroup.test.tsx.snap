// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioGroup should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="radio__group flex flex-col"
    data-testid="arke-radio-group"
  >
    <div
      class="radio__container w-fit flex flex-row-reverse items-center gap-2"
    >
      <div
        class="radio radio--secondary"
        data-testid="arke-radio"
      >
        <input
          class="opacity-0"
          name="test"
          type="radio"
          value="option1"
        />
      </div>
    </div>
    <div
      class="radio__container w-fit flex flex-row-reverse items-center gap-2"
    >
      <div
        class="radio radio--secondary"
        data-testid="arke-radio"
      >
        <input
          class="opacity-0"
          name="test"
          type="radio"
          value="option2"
        />
      </div>
    </div>
    <div
      class="radio__container w-fit flex flex-row-reverse items-center gap-2"
    >
      <div
        class="radio radio--secondary"
        data-testid="arke-radio"
      >
        <input
          class="opacity-0"
          name="test"
          type="radio"
          value="option3"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
